### PR TITLE
Fix KeyAvailable and ReadKey() in ConsolePal.Windows.

### LINKED ISSUE #102425: Bugs in `Console` members `KeyAvailable` and `ReadKey()` on Windows platform
## General

The `public static System.Console` API mostly redirects to the platform-specific implementations in `internal static System.ConsolePal`. For Windows, that's in [ConsolePal.Windows.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Console/src/System/ConsolePal.Windows.cs), so I will refer to the latter directly.

https://github.com/dotnet/runtime/blob/5474ab5c50fc639e780e0322cd500b2e43ee9b0d/src/libraries/System.Console/src/System/ConsolePal.Windows.cs#L284

https://github.com/dotnet/runtime/blob/5474ab5c50fc639e780e0322cd500b2e43ee9b0d/src/libraries/System.Console/src/System/ConsolePal.Windows.cs#L323

## Bug 1: Inconsistent logic between `KeyAvailable` and `ReadKey()`

The code peeking at the input queue in `KeyAvailable` uses a different logic than the code reading the input queue in `ReadKey()`. As a result, when invoking `KeyAvailable`,

- An unicode char synthesized via a `Alt` key-up event will be silently eaten and is no longer available to a subsequent `ReadKey()` call. OTOH, `ReadKey()` without previous `KeyAvailable` will read that char correctly.\
This affects input via an `Alt` plus `numpad number` sequence, pasted chars without key equivalent, and problably IME-generated chars as well.

- A keydown event in the input queue for any combination of `Alt` with `numpad number` keys, `arrow` keys, `page up/down` keys, `Home/End` keys, and the `Insert` key will return `true`, signaling that a subsequent call to `ReadKey()` should return that key and **should not block**.\
Unfortunately, the opposite happens, because `ReadKey()` silently eats these keys in anticipation of a subsequent synthesized key (which will only happen for some of these, see below). As a result, a call to `ReadKey()` **will block and not return**, despite `KeyAvailable` having signaled otherwise.

## Bug 2: `ReadKey()` incorrectly skips valid keydown events

This is again related to synthesized unicode chars from an `Alt` plus `numpad number` input sequence, which are surfaced via a subsequent `Alt` key-up event.

The code correctly skips such keydown events for virtual `numpad number` keys. Each of the corresponding physical keys has a second virtual key that gets sent when `NumLock` is off, so `ReadKey()` skips them as well, because synthesizing happens regardless of the NumLock state.

However, on most keyboards, those keys exist twice - once on the numpad, and once on the arrow pad and control pad. **Only the keys on the numpad** will produce a synthesized unicode char, but `ReadKey()` lumps them together causing the keys on the control and arrow pad to be silently eaten.

## Proposed fix

https://github.com/dotnet/runtime/compare/main...mawosoft:runtime:fix-consolepal-windows

I can submit a PR if that's desirable.

### PR DESCRIPTION
**Pull request by mawosoft**
Contributes to #102425.

### CODE REVIEW

**Comment by mawosoft** [comment]
cc @adamsitnik
May warrant a backport?

File: src/libraries/System.Console/src/System/ConsolePal.Windows.cs [hunk 1]
```diff
@@ -200,19 +200,33 @@ internal static TextReader GetOrCreateReader()
...
-        private static bool IsModKey(Interop.INPUT_RECORD ir)
-        {
-            // We should also skip over Shift, Control, and Alt, as well as caps lock.
-            // Apparently we don't need to check for 0xA0 through 0xA5, which are keys like
-            // Left Control & Right Control. See the ConsoleKey enum for these values.
-            ushort keyCode = ir.keyEvent.wVirtualKeyCode;
-            return ((keyCode >= 0x10 && keyCode <= 0x12)
-                    || keyCode == 0x14 || keyCode == 0x90 || keyCode == 0x91);
+            if (ir.EventType != Interop.KEY_EVENT) return false;
+            if (ir.keyEvent.bKeyDown == Interop.BOOL.FALSE)
```
Comment (adamsitnik) [inline src/libraries/System.Console/src/System/ConsolePal.Windows.cs:206]: From https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md: *Braces may be omitted only if the body of every block associated with an if/else if/.../else compound statement is placed on a single line.*

Please add braces to every if/else block in this method.

```suggestion
            if (ir.EventType != Interop.KEY_EVENT)
            {
                return false;
            }
            else if (ir.keyEvent.bKeyDown == Interop.BOOL.FALSE)
```

File: src/libraries/System.Console/src/System/ConsolePal.Windows.cs [hunk 2]
```diff
@@ -200,19 +200,33 @@ internal static TextReader GetOrCreateReader()
...
+            {
+                return ir.keyEvent.wVirtualKeyCode == AltVKCode && ir.keyEvent.uChar != 0;
```
Comment (adamsitnik) [inline src/libraries/System.Console/src/System/ConsolePal.Windows.cs:208]: You have removed most of the comments and added very few. Could you please comment what each `if/else` block is doing?

**Review by adamsitnik — COMMENTED** [review]
@mawosoft big thanks for your contribution! Could you please add the missing comments? 

In the meantime, I am going to buy a new battery for my numpad keyboard so I can test all scenarios manually.

Thank you!

**Review by adamsitnik — COMMENTED** [review]
@mawosoft big thanks for addressing my feedback! I bought the battery for my numeric keyboard and was able to test the fix.

I've used following app for testing:

```cs
Console.OutputEncoding = System.Text.Encoding.UTF8;
Console.WriteLine($"Output encoding is: {Console.OutputEncoding}");

bool callKeyAvailableFirst = args.Length > 0;
if (callKeyAvailableFirst) Console.WriteLine("The app will call Console.KeyAvailable first.");

ConsoleKeyInfo info = default;
do
{
    if (callKeyAvailableFirst)
    {
        while (!Console.KeyAvailable) ;
    }

    info = Console.ReadKey(intercept: true);
    Console.WriteLine($"Got: key: {info.Key}, mod: {info.Modifiers}, char: {info.KeyChar}");
} while (info.Key != ConsoleKey.Escape);

```

I've built your fork and was running the app using corerun:

```cmd
git clone https://github.com/mawosoft/runtime.git --branch fix-consolepal-windows fix-consolepal-windows
# skipped for brevity
D:\projects\forks\fix-consolepal-windows\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\corerun.exe .\bin\Release\net8.0\KeyTests.dll keyAvailable
```

> Bug 1: Inconsistent logic between KeyAvailable and ReadKey
> This affects input via an Alt plus numpad number sequence
> Unfortunately, the opposite happens, because ReadKey() silently eats these keys in anticipation of a subsequent synthesized

I've tried the following combination: "Alt+NumPad1" and the outcome is the same (nothing gets printed). @mawosoft am I missing something?

> However, on most keyboards, those keys exist twice - once on the numpad, and once on the arrow pad and control pad.

Do you mean the Home, PageUp and others?

I've tried the following combination: "Alt+PageUp" (not form the numpad) and with your fix the output gets printed, without it, it does not 👍

**Comment by mawosoft** [comment]
> > Bug 1: Inconsistent logic between KeyAvailable and ReadKey
> > This affects input via an Alt plus numpad number sequence
> > Unfortunately, the opposite happens, because ReadKey() silently eats these keys in anticipation of a subsequent synthesized
> 
> I've tried the following combination: "Alt+NumPad1" and the outcome is the same (nothing gets printed). @mawosoft am I missing something?

With the new version and for `Alt down, Num1 down, Num1 up, Alt up`, you should get: `key: 18, mod: None, char: ☺`, (which is codepoint 0x263a) after releasing the Alt key. Should also be independent of the NumLock state.

I didn't test it with this build, but I'm using almost the same code in a proprietary app, and there both automated and manual tests show the expected result.

> 
> > However, on most keyboards, those keys exist twice - once on the numpad, and once on the arrow pad and control pad.
> 
> Do you mean the Home, PageUp and others?

Yep. I've listed them in the comments now: Insert, PageUp/Down, Home/End, arrow keys.

**Comment by mawosoft** [comment]
@adamsitnik Afterthought: You probably know this already, but just in case...

If you run this in some kind of terminal app, e.g. the terminal inside Visual Studio, you may get different results depending on how the terminal app translates those keys.

**Comment by adamsitnik** [comment]
> If you run this in some kind of terminal a

I am using Windows Terminal on Windows 11 x64

> With the new version and for `Alt down, Num1 down, Num1 up, Alt up`, you should get: `key: 18, mod: None, char: ☺`, (which is codepoint 0x263a) after releasing the Alt key. Should also be independent of the NumLock state.

I've tried that with no luck:
- Press Alt
- Press Num1
- Release Num1
- Release Alt

> I didn't test it with this build,

Is there any chance you could try it locally? Your PR is overall an improvement, but it would be great co clarify that before merging.

Here you can find what is required to build dotnet/runtime: https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/windows-requirements.md

```cmd
git clone https://github.com/mawosoft/runtime.git --branch fix-consolepal-windows fix-consolepal-windows
cd fix-consolepal-windows
.\build.cmd -c Release -subset clr+libs+libs.tests
```

This should produce `corerun.exe` which just runs the code from the fist argument that you provide (a path to a `.dll` with test app).

```cmd
.\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\corerun.exe $pathToTestApp.dll
```

Corerun is going to use the local build of dotnet/runtime.

If you apply some changes to `System.Console`, you need to run following command to get the files updated:

```cmd
.\dotnet.cmd build .\src\libraries\System.Console\System.Console.sln -c Release
```

> there both automated and manual tests show the expected result.

May I ask how have you implemented automated tests for it? Have you just mocked the input for the `IsReadKeyEvent` method?

**Comment by mawosoft** [comment]
> > If you run this in some kind of terminal a
> 
> I am using Windows Terminal on Windows 11 x64

Can you run this in a regular ConHost? I should note that I did all testing on Win10 x64. Not sure if Win11 could throw a wrench into things.

Anyway, with Windows Terminal or any other you basically get kbd input --> translate to virtual key sequence -> send VT sequence -> read and interpret VT sequence -> synthesize kbd events, which may not match the original input at all.

> Is there any chance you could try it locally? Your PR is overall an improvement, but it would be great co clarify that before merging.

Not before Wednesday. I guess we have to postpone it then, as you suggested initially.

> May I ask how have you implemented automated tests for it? Have you just mocked the input for the `IsReadKeyEvent` method?

No, I used the native SendInput (user32) to run through most possible key combos in Conhost. It's a custom impl, not AppDriver or something. It's a bit tricky to sync properly, and while automated, I only ran them locally, not in a CI environment.

**Review by adamsitnik — APPROVED** [review]
> Can you run this in a regular ConHost?

I did. It's the same. FWIW the following condition is true and `false` is being returned, but there is no follow up event that this comment mentions:

https://github.com/dotnet/runtime/blob/05240f13a463a192e88cc90b9493a36cca3ecb0d/src/libraries/System.Console/src/System/ConsolePal.Windows.cs#L237-L243

@mawosoft I've spent more time testing your code. It solves one of the two described problems and does not seem to introduce any new ones, so I believe it's worth merging now. If you find some free time and a way to fix the other issue, please send a follow up PR.

**Comment by mawosoft** [comment]
@adamsitnik Thanks for the tests. 

If the follow-up event (the Alt keyup with the unicode char) doesn't appear, it indicates something driver-related. Is your numeric keyboard by any chance just a separate number block used in addition to the default keyboard? Do the Alt Numpad keys produce 'special' chars in other apps, e.g. in a plain old cmd.exe?

An alternative would be to use the On-Screen Keyboard where you can enable the numpad section regardless of your physical keyboard layout.

**Comment by adamsitnik** [comment]
> Is your numeric keyboard by any chance just a separate number block used in addition to the default keyboard

It is. To be exact it's Microsoft Sculpt Ergonomic Desktop

image

> Do the Alt Numpad keys produce 'special' chars in other apps, e.g. in a plain old cmd.exe?

They don't.

> An alternative would be to use the On-Screen Keyboard where you can enable the numpad section regardless of your physical keyboard layout.

I've tried that and the effect was the same. (to be exact I've enabled NumLock in the options, then started the app, pressed the right alt and then selected 1 from NumPad).

image

**Pull request merged**